Add explicit ddc/capability LR/SC

arichardson opened the pull request:
Untested but should hopefully work.

arichardson committed:
[CHERI-RISCV] Implement explicit DDC/capability load-reserved

These instructions aren't implemented in the sail mode so I missed them
when adding initial atomics support.

arichardson committed:
[CHERI-RISCV] Implement explicit DDC/capability store-conditional

These instructions aren't implemented in the sail mode so I missed them
when adding initial atomics support.

jrtc27 commented on target/riscv/insn32-cheri-32.decode:
```diff
@@ -44,6 +49,6 @@ sccap       1111100  ..... ..... 000 01011 1011011 @r_2source  # This is sdcap (
-sc_cap      00011 . . ..... ..... 011 ..... 0101111 @atom_st
+lr_c        00010 . . 00000 ..... 011 ..... 0101111 @atom_ld
+sc_c        00011 . . ..... ..... 011 ..... 0101111 @atom_st
 amoswap_cap 00001 . . ..... ..... 011 ..... 0101111 @atom_st
```
```suggestion
amoswap_c   00001 . . ..... ..... 011 ..... 0101111 @atom_st
```

jrtc27 commented on target/riscv/insn32-cheri-64.decode:
```diff
@@ -41,16 +41,25 @@ lcddc       1111101  10111 ..... 000 ..... 1011011 @r2  # Would be lqddc for RV1
+
+sc_d_ddc    1111100  ..... ..... 000 10011 1011011 @atom_st_cap_or_ddc  # RV64-only (sc_c_ddc for RV32)
+sc_d_cap    1111100  ..... ..... 000 11011 1011011 @atom_st_cap_or_ddc  # RV64-only (sc_c_cap for RV32)
+sc_c_ddc    1111100  ..... ..... 000 10100 1011011 @atom_st_cap_or_ddc  # RV64-only (would be sc_q_ddc for RV128)
```
The inconsistency between the sorting of loads and stores isn't great, but it's what you've already done for the explicit non-atomic loads and stores in this file... I prefer the sorting used for the loads, as it's also the order the encodings are allocated in (and what you've done for the stores even in `insn32-cheri.decode`).

jrtc27 commented on target/riscv/insn32-cheri-64.decode:
```diff
@@ -41,16 +41,25 @@ lcddc       1111101  10111 ..... 000 ..... 1011011 @r2  # Would be lqddc for RV1
-sc_cap      00011 . . ..... ..... 100 ..... 0101111 @atom_st
+lr_c        00010 . . 00000 ..... 100 ..... 0101111 @atom_ld
+sc_c        00011 . . ..... ..... 100 ..... 0101111 @atom_st
 amoswap_cap 00001 . . ..... ..... 100 ..... 0101111 @atom_st
```
```suggestion
amoswap_c   00001 . . ..... ..... 100 ..... 0101111 @atom_st
```

jrtc27 commented on target/riscv/insn32-cheri.decode:
```diff
@@ -96,6 +96,11 @@ cgetaddr    1111111  01111 ..... 000 ..... 1011011 @r2
+@atom_ld_cap_or_ddc .......   ..... ..... ... ..... ....... &atomic aq=1 rl=1 rs2=0  %rs1 %rd
+@atom_st_cap_or_ddc .......   ..... ..... ... ..... ....... &atomic aq=1 rl=1 rd=0   %rs2 %rs1
+
+
```
Remove this line

jrtc27 commented on target/riscv/insn_trans/trans_cheri.inc.c:
```diff
@@ -425,35 +425,75 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
+static inline bool trans_lr_c(DisasContext *ctx, arg_lr_c *a)
 {
-    /* In a parallel context, stop the world and single step.  */
+    // Note: The capmode dependent address interpretation happens in the
```
Do you think this is something you will be changing, to match the non-cap-width atomics?

arichardson replied:
I'm not sure. It might make sense to do so but I don't it reduces code duplication very much.
Doing it all with cmpxchg will be difficult so we might as well do it all in the helper with an EXCP_ATOMIC

jrtc27 commented on target/riscv/insn_trans/trans_cheri.inc.c:
```diff
@@ -425,35 +425,75 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
 
 // Atomic ops
-static inline bool trans_lr_cap(DisasContext *ctx, arg_lr_cap *a)
+static inline bool trans_lr_cap_impl(DisasContext *ctx, arg_atomic *a,
```
```suggestion
static inline bool trans_lr_c_impl(DisasContext *ctx, arg_atomic *a,
```

jrtc27 commented on target/riscv/insn_trans/trans_cheri.inc.c:
```diff
@@ -425,35 +425,75 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
+    return trans_lr_cap_impl(ctx, a, &gen_helper_lr_c_cap);
+}
+
+static inline bool trans_sc_cap_impl(DisasContext *ctx, arg_atomic *a,
```
```suggestion
static inline bool trans_sc_c_impl(DisasContext *ctx, arg_atomic *a,
```

jrtc27 commented on target/riscv/insn_trans/trans_cheri.inc.c:
```diff
@@ -425,35 +425,75 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
+    return trans_sc_cap_impl(ctx, a, &gen_helper_sc_c_cap);
+}
+
 static inline bool trans_amoswap_cap(DisasContext *ctx, arg_amoswap_cap *a)
```
```suggestion
static inline bool trans_amoswap_c(DisasContext *ctx, arg_amoswap_cap *a)
```

jrtc27 commented on target/riscv/insn_trans/trans_cheri.inc.c:
```diff
@@ -425,35 +425,75 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
+
 static inline bool trans_amoswap_cap(DisasContext *ctx, arg_amoswap_cap *a)
 {
     /* In a parallel context, stop the world and single step.  */
```
Comments should be updated like LR/SC.

arichardson committed:
[CHERI-RISCV] Improve consistency of capability amoswap naming

Co-Authored-By: Jessica Clarke <jrtc27@jrtc27.com>

jrtc27 commented on target/riscv/insn_trans/trans_cheri.inc.c:
```diff
@@ -425,36 +425,76 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
-    /* In a parallel context, stop the world and single step.  */
+    // Note: The capmode dependent address interpretation happens in the
+    // helper and not during translation.
+    return trans_lr_cap_impl(ctx, a, &gen_helper_lr_c_modedep);
```
These all need updating to match my suggested renamings

arichardson committed:
[CHERI-RISCV] Reorder and rename explicit load/store decoding

The order is now consistent with the encoding allocation. I also changed
the names to use ld_<size>_(ddc|cap) and st_<size>_(ddc|cap) to avoid
confusing names such as sccap which could be considered store-conditional.

arichardson commented:
ping?

arichardson committed:
[CHERI-RISCV] Add missing REQUIRE_EXT(RVA) for capability atomics

jrtc27 reviewed (approved):
Seems good assuming it works
